File: packages/ld-adapter/src/__tests__/launch-darkly-provider.test.ts
import { readFileSync } from 'node:fs';
import { dirname, join } from 'node:path';
import { fileURLToPath } from 'node:url';
import nock from 'nock';
import { afterEach, beforeEach, describe, expect, it } from 'vitest';
import LaunchDarklyApi from 'launchdarkly-api';
import { createLaunchDarklyProvider } from '../provider/launch-darkly-provider.js';
import { ApprovalRequiredError } from '../errors/ld-adapter-error.js';
import { SEMANTIC_PATCH_CONTENT_TYPE } from '../client/ld-api-client.js';

const baseUrl = 'https://app.launchdarkly.com';
const fixturesDir = join(dirname(fileURLToPath(import.meta.url)), 'fixtures');
const flagFixture = JSON.parse(
	readFileSync(join(fixturesDir, 'flag-boolean.json'), 'utf8'),
);
const patchSuccessFixture = JSON.parse(
	readFileSync(join(fixturesDir, 'patch-canary-success.json'), 'utf8'),
);
const patch422Fixture = JSON.parse(
	readFileSync(join(fixturesDir, 'patch-422-invalid.json'), 'utf8'),
);
const patch429Fixture = JSON.parse(
	readFileSync(join(fixturesDir, 'patch-429-retry.json'), 'utf8'),
);

describe('LaunchDarklyProvider nock integration', () => {
	beforeEach(() => {
		nock.cleanAll();
		const apiClient = LaunchDarklyApi.ApiClient.instance;
		apiClient.basePath = baseUrl;
		apiClient.authentications.ApiKey.apiKey = 'test-token';
		apiClient.defaultHeaders['LD-API-Version'] = '20240415';
	});

	afterEach(() => {
		nock.cleanAll();
	});

	it('PROV-01: reads flag state from GET fixture', async () => {
		nock(baseUrl)
			.get('/api/v2/flags/default/sample-feature')
			.reply(200, flagFixture);

		const provider = createLaunchDarklyProvider({
			accessToken: 'test-token',
			baseUrl,
		});
		const state = await provider.getFlagState({
			projectKey: 'default',
			flagKey: 'sample-feature',
			environmentKey: 'production',
		});

		expect(state.variations).toHaveLength(2);
		expect(state.environments.production?.on).toBe(true);
		expect(nock.isDone()).toBe(true);
	});

	it('PROV-02/03: semantic patch with resolved variation ids', async () => {
		let patchBody: Record<string, unknown> | undefined;

		nock(baseUrl)
			.get('/api/v2/flags/default/sample-feature')
			.times(2)
			.reply(200, flagFixture);

		nock(baseUrl)
			.patch('/api/v2/flags/default/sample-feature', (body) => {
				patchBody = body as Record<string, unknown>;
				return true;
			})
			.reply(200, patchSuccessFixture);

		const provider = createLaunchDarklyProvider({
			accessToken: 'test-token',
			baseUrl,
		});

		await provider.applyTargeting({
			projectKey: 'default',
			flagKey: 'sample-feature',
			intent: {
				environmentKey: 'production',
				rollout: {
					mode: 'fallthrough',
					treatmentVariationRef: { by: 'value', value: true },
					controlVariationRef: { by: 'value', value: false },
					treatmentPercentThousandths: 10_000,
					rolloutContextKind: 'user',
					rolloutBucketBy: 'user',
				},
			},
		});

		const instructions = (patchBody?.instructions ?? []) as Array<{
			kind: string;
			rolloutWeights?: Record<string, number>;
		}>;
		expect(instructions.some((i) => i.kind === 'turnFlagOn')).toBe(true);
		const rollout = instructions.find(
			(i) => i.kind === 'updateFallthroughVariationOrRollout',
		);
		expect(rollout?.rolloutWeights).toEqual({
			'var-on': 10_000,
			'var-off': 90_000,
		});
		expect(
			Object.values(rollout?.rolloutWeights ?? {}).reduce((a, b) => a + b, 0),
		).toBe(100_000);
	});

	it('422 returns error with single PATCH attempt', async () => {
		nock(baseUrl)
			.get('/api/v2/flags/default/sample-feature')
			.reply(200, flagFixture);
		const patchScope = nock(baseUrl)
			.patch('/api/v2/flags/default/sample-feature')
			.once()
			.reply(422, patch422Fixture);

		const provider = createLaunchDarklyProvider({
			accessToken: 'test-token',
			baseUrl,
		});

		await expect(
			provider.applyTargeting({
				projectKey: 'default',
				flagKey: 'sample-feature',
				intent: {
					environmentKey: 'production',
					turnOn: true,
				},
			}),
		).rejects.toBeTruthy();
		expect(patchScope.isDone()).toBe(true);
	});

	it('429 then 200 retries successfully', async () => {
		nock(baseUrl)
			.get('/api/v2/flags/default/sample-feature')
			.times(2)
			.reply(200, flagFixture);

		nock(baseUrl)
			.patch('/api/v2/flags/default/sample-feature')
			.reply(429, patch429Fixture, { 'Retry-After': '0' })
			.patch('/api/v2/flags/default/sample-feature')
			.reply(200, patchSuccessFixture);

		const provider = createLaunchDarklyProvider({
			accessToken: 'test-token',
			baseUrl,
		});

		await expect(
			provider.applyTargeting({
				projectKey: 'default',
				flagKey: 'sample-feature',
				intent: { environmentKey: 'production', turnOn: true },
			}),
		).resolves.toBeDefined();
	});

	it('405 maps to ApprovalRequiredError', async () => {
		nock(baseUrl)
			.get('/api/v2/flags/default/sample-feature')
			.reply(200, flagFixture);
		nock(baseUrl)
			.patch('/api/v2/flags/default/sample-feature')
			.reply(405, { message: 'approval required' });

		const provider = createLaunchDarklyProvider({
			accessToken: 'test-token',
			baseUrl,
		});

		await expect(
			provider.applyTargeting({
				projectKey: 'default',
				flagKey: 'sample-feature',
				intent: { environmentKey: 'production', turnOn: true },
			}),
		).rejects.toBeInstanceOf(ApprovalRequiredError);
	});

	it('uses semantic patch content type on PATCH requests', async () => {
		nock(baseUrl)
			.get('/api/v2/flags/default/sample-feature')
			.times(2)
			.reply(200, flagFixture);

		nock(baseUrl)
			.patch('/api/v2/flags/default/sample-feature')
			.reply(function (_uri, _body) {
				expect(this.req.headers['content-type']).toContain(
					SEMANTIC_PATCH_CONTENT_TYPE.split(';')[0],
				);
				expect(this.req.headers['content-type']).toContain(
					'launchdarkly.semanticpatch',
				);
				return [200, patchSuccessFixture];
			});

		const provider = createLaunchDarklyProvider({
			accessToken: 'test-token',
			baseUrl,
		});
		await provider.applyTargeting({
			projectKey: 'default',
			flagKey: 'sample-feature',
			intent: { environmentKey: 'production', turnOn: true },
		});
	});
});

File: kotlin/modules/ld-adapter/src/test/kotlin/com/ffpromo/ldadapter/provider/LaunchDarklyProviderTest.kt
package com.ffpromo.ldadapter.provider

import com.ffpromo.contracts.ApplyTargetingInput
import com.ffpromo.contracts.GetFlagStateInput
import com.ffpromo.contracts.LaunchDarklyClientConfig
import com.ffpromo.contracts.RolloutIntent
import com.ffpromo.contracts.TargetingIntent
import com.ffpromo.contracts.VariationRef
import com.ffpromo.ldadapter.client.SEMANTIC_PATCH_CONTENT_TYPE
import com.ffpromo.ldadapter.errors.ApprovalRequiredError
import kotlinx.coroutines.test.runTest
import kotlinx.serialization.json.Json
import kotlinx.serialization.json.JsonPrimitive
import okhttp3.mockwebserver.MockResponse
import okhttp3.mockwebserver.MockWebServer
import org.junit.jupiter.api.AfterEach
import org.junit.jupiter.api.Assertions.assertEquals
import org.junit.jupiter.api.Assertions.assertTrue
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.assertThrows

class LaunchDarklyProviderTest {
    private lateinit var server: MockWebServer
    private val json = Json { ignoreUnknownKeys = true }

    @BeforeEach
    fun setUp() {
        server = MockWebServer()
        server.start()
    }

    @AfterEach
    fun tearDown() {
        server.shutdown()
    }

    private fun fixture(name: String): String =
        javaClass.getResource("/fixtures/$name")!!.readText()

    private fun provider(): LaunchDarklyProvider =
        createLaunchDarklyProvider(
            LaunchDarklyClientConfig(
                accessToken = "test-token",
                baseUrl = server.url("/").toString().trimEnd('/'),
            ),
        )

    @Test
    fun `PROV-01 reads flag state from GET fixture`() = runTest {
        server.enqueue(MockResponse().setResponseCode(200).setBody(fixture("flag-boolean.json")))

        val state = provider().getFlagState(
            GetFlagStateInput(
                projectKey = "default",
                flagKey = "sample-feature",
                environmentKey = "production",
            ),
        )

        assertEquals(2, state.variations.size)
        assertEquals(true, state.environments["production"]?.on)
        assertEquals(1, server.requestCount)
    }

    @Test
    fun `PROV-02 and PROV-03 semantic patch with resolved variation ids`() = runTest {
        val flagFixture = fixture("flag-boolean.json")
        server.enqueue(MockResponse().setResponseCode(200).setBody(flagFixture))
        server.enqueue(MockResponse().setResponseCode(200).setBody(fixture("patch-canary-success.json")))
        server.enqueue(MockResponse().setResponseCode(200).setBody(flagFixture))

        provider().applyTargeting(
            ApplyTargetingInput(
                projectKey = "default",
                flagKey = "sample-feature",
                intent = TargetingIntent(
                    environmentKey = "production",
                    rollout = RolloutIntent(
                        mode = "fallthrough",
                        treatmentVariationRef = VariationRef.ByValue(JsonPrimitive(true)),
                        controlVariationRef = VariationRef.ByValue(JsonPrimitive(false)),
                        treatmentPercentThousandths = 10_000,
                        rolloutContextKind = "user",
                        rolloutBucketBy = "user",
                    ),
                ),
            ),
        )

        server.takeRequest()
        val patchBody = server.takeRequest().body.readUtf8()
        assertTrue(patchBody.contains("turnFlagOn"))
        assertTrue(patchBody.contains("updateFallthroughVariationOrRollout"))
        assertTrue(patchBody.contains("\"var-on\":10000"))
        assertTrue(patchBody.contains("\"var-off\":90000"))
    }

    @Test
    fun `422 returns error with single PATCH attempt`() = runTest {
        server.enqueue(MockResponse().setResponseCode(200).setBody(fixture("flag-boolean.json")))
        server.enqueue(MockResponse().setResponseCode(422).setBody(fixture("patch-422-invalid.json")))

        assertThrows<Throwable> {
            provider().applyTargeting(
                ApplyTargetingInput(
                    projectKey = "default",
                    flagKey = "sample-feature",
                    intent = TargetingIntent(environmentKey = "production", turnOn = true),
                ),
            )
        }

        server.takeRequest()
        server.takeRequest()
        assertEquals(2, server.requestCount)
    }

    @Test
    fun `429 then 200 retries successfully`() = runTest {
        val flagFixture = fixture("flag-boolean.json")
        server.enqueue(MockResponse().setResponseCode(200).setBody(flagFixture))
        server.enqueue(
            MockResponse()
                .setResponseCode(429)
                .setBody(fixture("patch-429-retry.json"))
                .addHeader("Retry-After", "0"),
        )
        server.enqueue(MockResponse().setResponseCode(200).setBody(fixture("patch-canary-success.json")))
        server.enqueue(MockResponse().setResponseCode(200).setBody(flagFixture))

        provider().applyTargeting(
            ApplyTargetingInput(
                projectKey = "default",
                flagKey = "sample-feature",
                intent = TargetingIntent(environmentKey = "production", turnOn = true),
            ),
        )

        assertEquals(4, server.requestCount)
    }

    @Test
    fun `405 maps to ApprovalRequiredError`() = runTest {
        server.enqueue(MockResponse().setResponseCode(200).setBody(fixture("flag-boolean.json")))
        server.enqueue(
            MockResponse()
                .setResponseCode(405)
                .setBody("""{"message":"approval required"}"""),
        )

        assertThrows<ApprovalRequiredError> {
            provider().applyTargeting(
                ApplyTargetingInput(
                    projectKey = "default",
                    flagKey = "sample-feature",
                    intent = TargetingIntent(environmentKey = "production", turnOn = true),
                ),
            )
        }
    }

    @Test
    fun `uses semantic patch content type on PATCH requests`() = runTest {
        val flagFixture = fixture("flag-boolean.json")
        server.enqueue(MockResponse().setResponseCode(200).setBody(flagFixture))
        server.enqueue(MockResponse().setResponseCode(200).setBody(fixture("patch-canary-success.json")))
        server.enqueue(MockResponse().setResponseCode(200).setBody(flagFixture))

        provider().applyTargeting(
            ApplyTargetingInput(
                projectKey = "default",
                flagKey = "sample-feature",
                intent = TargetingIntent(environmentKey = "production", turnOn = true),
            ),
        )

        server.takeRequest()
        val patchRequest = server.takeRequest()
        assertTrue(patchRequest.getHeader("Content-Type")!!.contains(SEMANTIC_PATCH_CONTENT_TYPE.split(';')[0]))
        assertTrue(patchRequest.getHeader("Content-Type")!!.contains("launchdarkly.semanticpatch"))
    }
}
